### ISSUE TITLE
Checker callbacks only triggered for some statements

### ISSUE THREAD

**Issue by T-Gruber** [llvm/llvm-project#94940]
I am currently working on different checkers for the Clang Static Analyser (llvm-project version release 17.x). One of the goals is to gain information about inline assembly code (GCCAsmStmt, ...). For this I wanted to use the callback method checkPreStmt<clang::GCCAsmStmt>. I noticed that not all stmts trigger the checkPreStmt. I am aware of this for statements that affect branches, such as IfStmt or similar, as the checkBranchCondition is triggered for these. However, I would expect the checkPreStmt/checkPostStmt to be called for statements such as clang::GCCAsmStmt or clang::CompoundStmt. 

Here is the code relevant code snippet of the implemented checker:
```cpp
class AsmChecker
    : public clang::ento::Checker<clang::ento::check::PreStmt<clang::Stmt> {

public:
  void checkPreStmt(const clang::Stmt *PreS,
                    clang::ento::CheckerContext &CC) const {
    llvm::outs() << "checkPreStmt: StmtClassName = " << PreS->getStmtClassName() << "\n";
  }
```

For the following example:
```cpp
extern int a, b, res;
int main() {
  __asm__("addl %%ebx, %%eax;" : "=a"(res) : "a"(a), "b"(b));
  return 0;
}
```

The corresponding checkPreStmt is not triggered for all Stmts contained in the AST:
```
checkPreStmt: StmtClassName = ImplicitCastExpr
checkPreStmt: StmtClassName = ImplicitCastExpr
checkPreStmt: StmtClassName = IntegerLiteral
checkPreStmt: StmtClassName = ReturnStmt
```

A look at the ExprEngine shows that ExprEngine::VisitGCCAsmStmt is called for GCCAsmStmts, but this function does not call the corresponding pre- and post-checkers (runCheckersForPreStmt/runCheckersForPostStmt) as is the case for some statements (e.g. VisitArraySubscriptExpr).

Is there a good reason for triggering the checkers only for certain statements or is this a current limitation?

I would be very grateful for any hints

**Comment by steakhal** [comment]
> Is there a good reason for triggering the checkers only for certain statements or is this a current limitation?

This is a bug. Thanks for noticing. Assign me for review if you plan to upstream a fix.

**Comment by T-Gruber** [comment]
Thank you for your quick feedback. I would like to work on a fix. There are a few things I would like to clarify first. 

## GCCAstmStmt
The ExprEngine::Visit function is responsible for the calls to the corresponding callbacks of the checkers. In the specific case of the clang::GCCAsmStmt, I have made the following changes as a temporary fix:
```cpp
void ExprEngine::Visit(const Stmt *S, ExplodedNode *Pred,
                       ExplodedNodeSet &DstTop) {
  ...
  switch (S->getStmtClass()) {
    ...
    case Stmt::GCCAsmStmtClass: {
      // modified code
      Bldr.takeNodes(Pred);
      ExplodedNodeSet PreVisit;
      getCheckerManager().runCheckersForPreStmt(PreVisit, Pred, S, *this);
      ExplodedNodeSet PostVisit;
      for (ExplodedNode *const TmpNode : PreVisit)
        VisitGCCAsmStmt(cast<GCCAsmStmt>(S), TmpNode, PostVisit);
      getCheckerManager().runCheckersForPostStmt(Dst, PostVisit, S, *this);
      Bldr.addNodes(Dst);
      break;

      // original code 
      Bldr.takeNodes(Pred);
      VisitGCCAsmStmt(cast<GCCAsmStmt>(S), Pred, Dst);
      Bldr.addNodes(Dst);
      break;
  }
```
What do you think, would this be a possible solution?


## Other Stmts
For many Stmts the fallthrough mechanism is used, for example the ImplicitCastExpr will call the VisitCast function. However, for some statements (e.g. clang::CompoundStmt), the following comment is provided in the ExprEngine::Visit:
```cpp
  // Cases that should never be evaluated simply because they shouldn't
  // appear in the CFG.
  ...
  case Stmt::CompoundStmtClass:
  ...
```
This explains why no callback function can be triggered for these statements. However, it is not entirely clear to me why this is explicitly implemented in this way.

**Comment by steakhal** [comment]
To unblock you, you would need a patch like this:
```diff
diff --git a/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp b/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
index b331be8f5664..8beb4cb49f21 100644
--- a/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
+++ b/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
@@ -2057,12 +2057,16 @@ void ExprEngine::Visit(const Stmt *S, ExplodedNode *Pred,
       llvm_unreachable("Support for MatrixSubscriptExpr is not implemented.");
       break;
 
-    case Stmt::GCCAsmStmtClass:
+    case Stmt::GCCAsmStmtClass: {
       Bldr.takeNodes(Pred);
-      VisitGCCAsmStmt(cast<GCCAsmStmt>(S), Pred, Dst);
+      ExplodedNodeSet PreVisit;
+      getCheckerManager().runCheckersForPreStmt(PreVisit, Pred, S, *this);
+      ExplodedNodeSet PostVisit;
+      for (ExplodedNode *N : PreVisit)
+        VisitGCCAsmStmt(cast<GCCAsmStmt>(S), N, PostVisit);
+      getCheckerManager().runCheckersForPostStmt(Dst, PostVisit, S, *this);
       Bldr.addNodes(Dst);
-      break;
-
+    } break;
     case Stmt::MSAsmStmtClass:
       Bldr.takeNodes(Pred);
       VisitMSAsmStmt(cast<MSAsmStmt>(S), Pred, Dst);
```

But this appears to be a more widespread issue. For example: `MSAsmStmtClass` does also not trigger a pre-post stmt callback.

`LambdaExpr`, `ReturnStmt` appears to only invoke the post-stmt, but not the pre-stmt. And I only spent like 2 minutes looking for these. I guess, we have some homework.

**Comment by T-Gruber** [comment]
Hi @steakhal, thanks for your answer. I just created a PR: https://github.com/llvm/llvm-project/pull/95409.
It only contains the changes regarding the GccAsmStmt. Is there a need to implement this in the same way for other statements?

**Comment by steakhal** [comment]
> Hi @steakhal, thanks for your answer. I just created a PR: https://github.com/llvm/llvm-project/pull/95409.
> It only contains the changes regarding the GccAsmStmt. Is there a need to implement this in the same way for other statements?

We didn't notice this so far, so I dont think we should rush with the fix.

So, unless you want to separately fix others, you dont need to. We probably accept such fixes.

**Comment by T-Gruber** [comment]
In addition to the GCCAsmStmt mentioned above, I also came across missing callbacks for the SwitchStmt. As mentioned here (https://lists.llvm.org/pipermail/cfe-dev/2013-December/033891.html), checkBranchCondition is only triggered for boolean branches and not for SwitchStmts.
This means that no callbacks can be used with SwitchStmts. If there is nothing fundamental against it, I would like to add the PreStmt and PostStmt calls in another PR.
What do you think?
